crypto/cms/cms_smime.c

```diff
@@ -313,26 +320,42 @@ int CMS_verify(CMS_ContentInfo *cms, STACK_OF(X509) *certs,
 
     /* Attempt to verify all signers certs */
 
-    if (!(flags & CMS_NO_SIGNER_CERT_VERIFY)) {
+    if ((flags & CMS_NO_SIGNER_CERT_VERIFY) == 0 || cadesVerify) {
+        if (cadesVerify) {
+            /* Certificate trust chain is required to check CAdES signature */
+            si_chains = OPENSSL_zalloc(scount * sizeof(si_chains[0]));
+            if (si_chains == NULL) {
+                CMSerr(CMS_F_CMS_VERIFY, ERR_R_MALLOC_FAILURE);
+                goto err;
+            }
+        }
         cms_certs = CMS_get1_certs(cms);
         if (!(flags & CMS_NOCRL))
             crls = CMS_get1_crls(cms);
         for (i = 0; i < sk_CMS_SignerInfo_num(sinfos); i++) {
```

Comment:
Optional: at this point `scount == sk_CMS_SignerInfo_num(sinfos)` so you could use scount here to optimize.